streams/src/main/java/org/apache/kafka/streams/kstream/internals/SourceSinkNode.java

```diff
@@ -0,0 +1,187 @@
+import org.apache.kafka.streams.processor.ProcessorSupplier;
+
+import java.util.Objects;
+
+/**
+ * Used for through nodes and other cases where you need
+ * a source node immediately followed by a sink node
+ */
+class SourceSinkNode extends StreamsGraphNode {
+
+    final Serde keySerde;
+    final Serde valueSerde;
+    final ChangedDeserializer changedDeserializer;
+    final ChangedSerializer changedSerializer;
+    final String sinkTopic;
+    final String sinkName;
+    final String funcOrFilterName;
+    final String sourceName;
+    final ProcessorSupplier processorSupplier;
+
+
+    public SourceSinkNode(final TopologyNodeType topologyNodeType,
+                          final Serde keySerde,
+                          final Serde valueSerde,
+                          final String sinkTopic,
+                          final String sinkName,
+                          final String funcOrFilterName,
+                          final String sourceName,
+                          final ProcessorSupplier processorSupplier,
+                          final ChangedDeserializer changedDeserializer,
+                          final ChangedSerializer changedSerializer) {
+        super(topologyNodeType);
+        this.keySerde = keySerde;
+        this.valueSerde = valueSerde;
+        this.sinkTopic = sinkTopic;
+        this.sinkName = sinkName;
+        this.funcOrFilterName = funcOrFilterName;
+        this.sourceName = sourceName;
+        this.processorSupplier = processorSupplier;
+        this.changedDeserializer = changedDeserializer;
+        this.changedSerializer = changedSerializer;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+
+    public static final class Builder {
+
+        String name;
+        Serde keySerde;
+        Serde valueSerde;
+        String sinkTopic;
+        String sinkName;
+        String filterName;
+        String sourceName;
+        TopologyNodeType topologyNodeType;
+        ProcessorSupplier processorSupplier;
+        ProcessDetails processDetails;
+        ChangedSerializer changedSerializer;
+        ChangedDeserializer changedDeserializer;
+
+        private Builder() {
+        }
+
+        public static Builder builder() {
+            return new Builder();
+        }
+
+        public Builder withKeySerde(Serde keySerde) {
+            this.keySerde = keySerde;
+            return this;
+        }
+
+        public Builder withValueSerde(Serde valueSerde) {
+            this.valueSerde = valueSerde;
+            return this;
+        }
+
+        public Builder withSinkTopic(String repartitionTopic) {
+            this.sinkTopic = repartitionTopic;
+            return this;
+        }
+
+        public Builder withSinkName(String sinkName) {
+            this.sinkName = sinkName;
+            return this;
+        }
+
+        public Builder withFilterName(String filterName) {
+            this.filterName = filterName;
```

Comment:
I'm not sure I understand what does this `filterName` used for, could you explain? It seems to be just used as the function name.

Reply:
It represents the name of the `ProcessorSupplier` used in `KStreamImpl.createRepartitionedSource`.  

The `filterName` eventually becomes the `funcOrFilterName` variable on the `SourceSinkNode`.  This name was chosen as in `KStreamImpl.createRepartitionedSource` we use a `filterName` for the `ProcessorSupplier`, but in `KGroupedTableImpl.buildAggregate` (again creating a repartitioned topic) we use `funcName` to name the `ProcessorSupplier`.  

The name chosen it too specific and should be something like `processorSupplierName` and the method on the builder should be `withProcessorSupplierName` or someting along those lines.

Additionally, the `SourceSinkNode` is used represent a repartitioned source, but the name will change to `RepartitionedGraphNode` as suggested.